pkg/internal/itest/direct_fetch_test.go

```diff
@@ -0,0 +1,181 @@
+package itest
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"math/rand"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lassie/pkg/internal/itest/mocknet"
+	"github.com/filecoin-project/lassie/pkg/internal/itest/testpeer"
+	"github.com/filecoin-project/lassie/pkg/internal/itest/unixfs"
+	"github.com/filecoin-project/lassie/pkg/internal/lp2ptransports"
+	"github.com/filecoin-project/lassie/pkg/lassie"
+	"github.com/filecoin-project/lassie/pkg/retriever"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-graphsync/storeutil"
+	bsnet "github.com/ipfs/go-libipfs/bitswap/network"
+	"github.com/ipfs/go-libipfs/bitswap/server"
+	"github.com/ipfs/go-unixfsnode"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/storage"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	host "github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	lpmock "github.com/libp2p/go-libp2p/p2p/net/mock"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	bitswapDirect    = 0
+	graphsyncDirect  = 1
+	transportsDirect = 2
+)
+
+func TestDirectFetch(t *testing.T) {
+
+	testCases := []struct {
+		name       string
+		directPeer int
+	}{
+		{
+			name:       "direct bitswap peer",
+			directPeer: bitswapDirect,
+		},
+		{
+			name:       "direct graphsync peer",
+			directPeer: graphsyncDirect,
+		},
+		{
+			name:       "peer responding on transports protocol",
+			directPeer: transportsDirect,
+		},
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			req := require.New(t)
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			rndSeed := time.Now().UTC().UnixNano()
+			t.Logf("random seed: %d", rndSeed)
+			var rndReader io.Reader = rand.New(rand.NewSource(rndSeed))
+
+			// build two files of 4MiB random bytes, packaged into unixfs DAGs
+			// (rootCid1 & rootCid2) and the original source data retained
+			// (srcData1, srcData2)
+			mrn := mocknet.NewMockRetrievalNet()
+
+			mrn.SetupNet(ctx, t)
+			mrn.SetupRetrieval(ctx, t)
+			graphsyncAddr := peer.AddrInfo{
+				ID:    mrn.HostRemote.ID(),
+				Addrs: mrn.HostRemote.Addrs(),
+			}
+			graphsyncMAs, err := peer.AddrInfoToP2pAddrs(&graphsyncAddr)
+			req.NoError(err)
+			rootCid, srcBytes := unixfs.GenerateFile(t, &mrn.LinkSystemRemote, rndReader, 4<<20)
+			srcData := []unixfs.DirEntry{{Path: "", Cid: rootCid, Content: srcBytes}}
+			qr := testQueryResponse
+			qr.MinPricePerByte = abi.NewTokenAmount(0) // make it free so it's not filtered
+			mrn.SetupQuery(ctx, t, rootCid, qr)
+			testPeerGenerator := testpeer.NewTestPeerGenerator(ctx, t, mrn.MN, []bsnet.NetOpt{}, []server.Option{})
```

Comment:
I think with https://github.com/filecoin-project/lassie/pull/116 this whole section becomes a bit more compact because mocknet can now do both

Reply:
not requesting you adapt, I'd be happy to adapt this if it gets merged first as it would be a good test of the work there since I haven't done a combined graphsync+bitswap test with it yet

Reply:
yea that'd be my preference -- merge, then adapt your PR as a test.